packages/react-reconciler/src/ReactFiberCacheComponent.new.js

```diff
@@ -0,0 +1,204 @@
+
+export function pushCacheProvider(workInProgress: Fiber, cache: Cache) {
+  if (!enableCache) {
+    return;
+  }
+  pushProvider(workInProgress, CacheContext, cache);
+}
+
+export function popCacheProvider(workInProgress: Fiber, cache: Cache) {
+  if (!enableCache) {
+    return;
+  }
+  popProvider(CacheContext, workInProgress);
+}
+
+export function requestCacheFromPool(renderLanes: Lanes): Cache {
+  if (!enableCache) {
+    return (null: any);
+  }
+  if (pooledCache !== null) {
+    return pooledCache;
+  }
+  // Create a fresh cache.
+  pooledCache = new Map();
+  return pooledCache;
+}
+
+export function pushRootCachePool(root: FiberRoot) {
+  if (!enableCache) {
+    return;
+  }
+  // When we start rendering a tree, read the pooled cache for this render
+  // from `root.pooledCache`. If it's currently `null`, we will lazily
+  // initialize it the first type it's requested. However, we only mutate
+  // the root itself during the complete/unwind phase of the HostRoot.
+  pooledCache = root.pooledCache;
+}
+
+export function popRootCachePool(root: FiberRoot, renderLanes: Lanes) {
+  if (!enableCache) {
+    return;
+  }
+  // The `pooledCache` variable points to the cache that was used for new
+  // cache boundaries during this render, if any. Stash it on the root so that
+  // parallel transitions may share the same cache. We will clear this field
+  // once all the transitions that depend on it (which we track with
+  // `pooledCacheLanes`) have committed.
+  root.pooledCache = pooledCache;
+  if (pooledCache !== null) {
+    root.pooledCacheLanes |= renderLanes;
+  }
+}
+
+export function restoreSpawnedCachePool(
+  offscreenWorkInProgress: Fiber,
+  prevCachePool: SpawnedCachePool,
+): SpawnedCachePool | null {
+  if (!enableCache) {
+    return (null: any);
+  }
+  const nextParentCache = isPrimaryRenderer
+    ? CacheContext._currentValue
+    : CacheContext._currentValue2;
+  if (nextParentCache !== prevCachePool.parent) {
+    // There was a refresh. Don't bother restoring anything since the refresh
+    // will override it.
+    return null;
+  } else {
+    // No refresh. Resume with the previous cache. This will override the cache
+    // pool so that any new Cache boundaries in the subtree use this one instead
+    // of requesting a fresh one.
+    push(prevFreshCacheOnStack, pooledCache, offscreenWorkInProgress);
+    pooledCache = prevCachePool.pool;
+
+    // Return the cache pool to signal that we did in fact push it. We will
+    // assign this to the field on the fiber so we know to pop the context.
+    return prevCachePool;
+  }
+}
+
+// Note: Ideally, `popCachePool` would return this value, and then we would pass
+// it to `getSuspendedCachePool`. But factoring reasons, those two functions are
+// in different phases/files. They are always called in sequence, though, so we
+// can stash the value here temporarily.
+let _suspendedPooledCache: Cache | null = null;
```

Comment:
What's the underscore about? :)

Reply:
It's a weird hack so I wanted to make it ugly